src/simple_http.rs

```diff
@@ -189,16 +248,40 @@ impl SimpleHttpTransport {
             return Err(Error::HttpErrorCode(response_code));
         }
 
-        let content_length = content_length.ok_or(Error::HttpParseError)?;
-
-        let mut buffer = vec![0; content_length];
+        // Read up to `content_length` bytes. Note that if there is no content-length
+        // header, we will assume an effectively infinite content length, i.e. we will
+        // just keep reading from the socket until it is closed.
+        let buffer = match content_length {
+            None => {
+                let mut buffer = Vec::with_capacity(INITIAL_RESP_ALLOC as usize);
+                // `take` consumes `reader` and drops it, unlocking the mutex
+                reader.take(FINAL_RESP_ALLOC).read_to_end(&mut buffer)?;
+                buffer
+            },
+            Some(n) if n > FINAL_RESP_ALLOC => {
+                return Err(Error::HttpResponseContentLengthTooLarge {
+                    length: n,
+                    max: FINAL_RESP_ALLOC,
+                });
+            },
+            Some(n) => {
+                let mut buffer = Vec::with_capacity(INITIAL_RESP_ALLOC as usize);
```

Comment:
Shouldn't the buffer have capacity `n`?

Reply:
Yeah, I think so. I went back and forth a couple times on this, but now that we bound `n` I think we should use it.